### PR TITLE
.NET Core compat

### PR DESCRIPTION
**Pull request by riezebosch**
Replaced the ApplicationException because it is not available in .NET Core and you shouldn't use it anyway
Restructured directory layout for projects due to incompatibility with xproj's
Replaced csproj with xproj with output for net35 and netstandard1.6
Updated nuspec to make package that contains support for net35 and .NET Core

Dit not update travis file so you have to build an package from the command line:  
`nuget
pack
.\src\RSG.Promise\C-Sharp-Promise.nuspec
-basepath
.\src\RSG.Promise\bin\Debug`

Kept your .NET 3.5 test project to prove the output is still working on that runtime.

### CODE REVIEW

**Comment by ashleydavis** [comment]
Thanks for your contribution.

How do I test your changes relating to xproj? I haven't used .Net core, can I use Visual Studio to build xproj?

Is it possible to have both csproj and xproj side-by-side?

Cheers
Ash

**Comment by riezebosch** [comment]
You'll need this: https://www.microsoft.com/net/core#windows (VS2015 Update 3 + .NET Core 1.0.1 VS2015 Tooling Preview 2).

Side by side is possisble but doesn't add much value because a) you'll have to duplicate the source files or include them as linked files in the csproj and b) the xproj already builds both assemblies for you when compiling.

**Comment by RoryDungan** [comment]
This seems to work fine although it breaks Travis CI so that would definitely need to be updated. Not sure if we really want to make everyone have to download the .NET Core tools in order to build for .NET Framework 3.5 though, so it would be good if there was a solution that didn't require that. Maybe once Visual Studio 2017 comes out this will no longer be an issue as .NET Core tools will be included in the main installer as opposed to an additional download, but until then just providing .xproj files isn't ideal.

Also apparently [.xproj files are being phased out](https://blogs.msdn.microsoft.com/dotnet/2016/07/15/net-core-roadmap/) and replaced with .csproj/MSBuild in the near future anyway so maybe it would be better to go down that route to support .NET Core?

**Comment by riezebosch** [comment]
Yeah, but there is an upgrade-path from xproj->csproj so this PR could come in handy.

**Comment by RoryDungan** [comment]
I think we'll wait till VS 2017 is out with built-in support for xproj, just because a lot of people who use this (RSG included) only require .NET 3.5 and having to download the .NET Core tools in order to build would make it inconvenient to use. 

.NET Core and .NET Framework 4.5+ also support [async and await](https://msdn.microsoft.com/en-us/library/mt674882.aspx) which is a good alternative to using promises.

**Comment by riezebosch** [comment]
The async/await is indeed where my journey started. I was looking for a way to turn good-ol'-events into a promise-like syntax. After creating my own framework on the fly I found out there was already something available, namely your package. But I couldn't use it in .NET Core. Therefor the port. It only then I came to the conclusion that events and future completion are disjunct principles. So I reverted back to good-ol'-events.

**Comment by riezebosch** [comment]
Crafted some new commits using the VS2017 SDK based projects and updated the travis information.

File: Promise_NonGeneric.cs [hunk 1]
```diff
@@ -434,8 +434,8 @@ private void AddProgressHandler(Action<float> onProgress, IRejectable rejectable
         /// </summary>
         private void InvokeRejectHandler(Action<Exception> callback, IRejectable rejectable, Exception value)
         {
-//            Argument.NotNull(() => callback);
-//            Argument.NotNull(() => rejectable);
+            //            Argument.NotNull(() => callback);
+            //            Argument.NotNull(() => rejectable);
```
Comment (riezebosch) [inline Promise_NonGeneric.cs]: I'm sorry CTRL+K,D messed up this comment. Though it was wrong in the first place.

**Comment by jdnichollsc** [comment]
Can I help to finish this pull request? :)

**Comment by RoryDungan** [comment]
That would be great! You'll just need to merge the latest changes to bring it up to speed with version 3.0 and fix the issues I mentioned before with the inconsistent version numbers and the missing strong name key. Also the new custom exception types here should be replaced with the ones from version 3 of the master.

Version 3.0 isn't a big change from 2.0, we just bumped the version number because it follows the JS standard for [`finally`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally) now, which is incompatible with our previous implementation, and following semver, any breaking change is a major version bump.

**Comment by RoryDungan** [comment]
Superseded by #81

**Pull request closed without merging**